[{"client_msg_id":"acc2ce87-7a89-4516-a92c-7413a3a06663","type":"message","user":"UDDCM376D","text":"自分のコードが悪い気がしてきた","ts":"1611292246.019400","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"h=GA"}]},{"client_msg_id":"0c0644e0-dca9-40f5-8ea8-72dfafe9469a","type":"message","user":"UDDCM376D","text":"むむ…","ts":"1611292219.018900","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"2Mie"}]},{"client_msg_id":"b745e6c7-214e-45ac-ab84-a5af9897b38c","type":"message","user":"UP6PA1UBE","text":"正常に catch されてるよう\n```messages clear\n\nlet s:V = vital#_vital#new()\nlet s:Promise = s:V.import('Async.Promise')\n\nfunction! s:sh(...) abort\n  echom 'sh' a:000\n  return s:Promise.new({rv, rj -\u0026gt; rv('foo')})\nendfunction\n\nfunction! s:make_response(...) abort\n  echom 'make_responsee' a:000\n  \" E492: Not an editor command\n  bar\nendfunction\n\nfunction! s:make_error_responsee(...) abort\n  echom 'make_error_responsee' a:000\nendfunction\n\nfunction! s:do(...) abort\n  return call('s:sh', a:000)\n        \\.then(function('s:make_response', ['baz']))\n        \\.catch(function('s:make_error_responsee'))\nendfunction\n\ncall s:do('qux')```\n`:messages`\n```sh ['qux']\nmake_responsee ['baz', 'foo']\nmake_error_responsee [{'throwpoint': 'function \u0026lt;SNR\u0026gt;267__worker[12]..\u0026lt;SNR\u0026gt;266__invoke_callback[6]..\u0026lt;SNR\u0026gt;246_make_response, 行 3', 'exception': 'Vim:E492: エディタのコマンドではありません:   bar'}]```","ts":"1611285711.018600","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"ycAW"}]},{"client_msg_id":"50c03086-967d-4745-bd86-3d3f65624808","type":"message","user":"UAZ33BKV2","text":"(便利に使わせていただいてます)","ts":"1611282188.017600","thread_ts":"1611281980.017500","subtype":"thread_broadcast","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"4Nixi"}],"root":{"client_msg_id":"2F26FA5F-038D-401B-854A-337B178CCA66","type":"message","user":"U776661HA","text":"（whiskyにAsync.Promise.Processってのがあるよ）","ts":"1611281980.017500","thread_ts":"1611281980.017500","reply_count":1,"team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"N5e"}]}},{"client_msg_id":"2F26FA5F-038D-401B-854A-337B178CCA66","type":"message","user":"U776661HA","text":"（whiskyにAsync.Promise.Processってのがあるよ）","ts":"1611281980.017500","thread_ts":"1611281980.017500","reply_count":1,"team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"N5e"}]},{"client_msg_id":"27273c51-ad09-49d0-98c4-7dd50625a188","type":"message","user":"UDDCM376D","text":"あとで、小さなサンプル書いて色々試してみます","ts":"1611277176.016500","edited":{"user":"UDDCM376D","ts":"1611277182.000000"},"team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"BfkCl"}]},{"client_msg_id":"b72e21a6-2f4f-4105-b09f-aa9897197195","type":"message","user":"UAZ33BKV2","text":"あー、そうか...追加は即座だし、さすがにthenが付いたけど、catchが付く前とかじゃないだろうしな...","ts":"1611276247.015900","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"uK9N"}]},{"client_msg_id":"ca073422-0f2d-4603-8e4b-95bdeaf6426b","type":"message","user":"UDDCM376D","text":"s:make_response 内で発生しているんで、thenやcatchの前に発生することはないと思います","ts":"1611276098.014800","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"a=P3i"}]},{"client_msg_id":"9fc5afd2-306e-42d8-889e-6f5dc2c82018","type":"message","user":"UAZ33BKV2","text":"タイミング的に `.then` や `.cactch` を追加する前に発生しているとかはない?\nエラーを能動的にtry catchして再スローしたら取れるとか?","ts":"1611274337.014200","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"lYf"}]},{"client_msg_id":"ca3424fc-46e0-4535-8d5e-f97a3b91dffb","type":"message","user":"UDDCM376D","text":"catch はしっかり書いているけど…","ts":"1611274128.013300","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"NX/DI"}]},{"client_msg_id":"51b0ba77-8135-42e2-904c-cc98b8e3cf5b","type":"message","user":"UDDCM376D","text":"PromiseIsHanlded が false のときに reject が発生すると Unhanlded Rejection になるのは分かったけど、\n↓こういうコードで、s:make_response 内で発生したエラーがcatchされないんすよね\n```function! s:sh(...) abort\n  let cmd = a:000\n  let stdout = ['']\n  let stderr = ['']\n\n  return s:Promise.new({\n        \\ rv, rj -\u0026gt; s:Job.start(cmd, {\n        \\   'on_stdout': function('s:on_receive', [stdout]),\n        \\   'on_stderr': function('s:on_receive', [stderr]),\n        \\   'on_exit': { e -\u0026gt;\n        \\     e ? rj(join(stderr, \"\\r\")) : rv(join(stdout, \"\\r\"))\n        \\   }\n        \\ })\n        \\})\nendfunction\n\n  return call('s:sh', cmd)\n        \\.then(function('s:make_response', [tmp_file]))\n        \\.catch(function('s:make_error_responsee'))```","ts":"1611274088.012700","edited":{"user":"UDDCM376D","ts":"1611274099.000000"},"team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"oQs"}]},{"client_msg_id":"a483b0d3-12d1-4ba6-a789-c65f94a20448","type":"message","user":"U03C4RC8X","text":"そしてあらためてvital Async.Promise ちゃんとしてる :tada:","ts":"1611273767.010600","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"g5sm"}]},{"client_msg_id":"69d0b3c7-9748-4aa9-97db-4073383fb9d6","type":"message","user":"U03C4RC8X","text":"おお、直感的だ \u0026gt; then, catch, finally, awaitでhandledがtrueに","ts":"1611273723.010100","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"VOl4F"}]},{"client_msg_id":"f842bd06-9d56-4ccd-acf8-1ac3527ae955","type":"message","user":"UDDCM376D","text":"難しい","ts":"1611273348.008800","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"erF"}]},{"client_msg_id":"bc681255-b873-4325-882b-fe275afe2b10","type":"message","user":"UDDCM376D","text":"なるほど？","ts":"1611272921.008400","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"a3O"}]},{"client_msg_id":"5d7c2cab-0e60-4f01-95a8-e5f72cea6598","type":"message","user":"UDDCM376D","text":"\u0026gt; 整理すると、*[[PromiseIsHandled]]がfalseのPromise*とは、*まだ`then`や`catch`, `finally`によってコールバック関数が登録されていないPromise*を指すことになります。\n\u0026gt; ここまでで分かったことは、Promiseオブジェクトの[[PromiseIsHandled]]は、そのPromiseオブジェクトに対して`then`, `catch`, `finally`が呼ばれるか、またはそのPromiseオブジェクトが`await`された場合にtrueになるフラグだということです。\n\u0026gt; そして、このフラグがfalseである（＝まだ`then`などが呼ばれていない）Promiseがrejectした（RejectPromise抽象操作が呼ばれた）場合にHostPromiseRejectionTrackerが呼ばれる、すなわちUnhandle Rejectionとして扱われるということです。\n","ts":"1611272919.008200","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"FfWGx"}]},{"client_msg_id":"2a26c387-03f0-4e07-99ae-85fd44092ba6","type":"message","user":"UDDCM376D","text":"\u003chttps://zenn.dev/uhyo/articles/unhandled-rejection-understanding\u003e","ts":"1611272529.007200","attachments":[{"fallback":"Zenn: PromiseのUnhandled Rejectionを完全に理解する","id":1,"title":"PromiseのUnhandled Rejectionを完全に理解する","title_link":"https://zenn.dev/uhyo/articles/unhandled-rejection-understanding","image_url":"https://res.cloudinary.com/dlhzyuewr/image/upload/s--ssJqbt1B--/g_south_west%2Ch_90%2Cl_fetch:aHR0cHM6Ly9zdG9yYWdlLmdvb2dsZWFwaXMuY29tL3plbm4tdXNlci11cGxvYWQvYXZhdGFyL2ljb25fNWNiMWUwMmU0ZC5qcGVn%2Cr_max%2Cw_90%2Cx_87%2Cy_72/co_rgb:222%2Cg_south_west%2Cl_text:notosansjp-medium.otf_37_bold:uhyo%2Cx_203%2Cy_98/c_fit%2Cco_rgb:222%2Cg_north_west%2Cl_text:notosansjp-medium.otf_70_bold:Promise%25E3%2581%25AEUnhandled%2520Rejection%25E3%2582%2592%25E5%25AE%258C%25E5%2585%25A8%25E3%2581%25AB%25E7%2590%2586%25E8%25A7%25A3%25E3%2581%2599%25E3%2582%258B%2Cw_1010%2Cx_90%2Cy_100/v1609308637/og/new_txlqub.png","blocks":null,"service_name":"Zenn","service_icon":"https://zenn.dev/images/icon.png","from_url":"https://zenn.dev/uhyo/articles/unhandled-rejection-understanding","original_url":"https://zenn.dev/uhyo/articles/unhandled-rejection-understanding"}],"team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"+QXB"}]},{"client_msg_id":"8acac3f1-0192-490f-bf66-ce3dd6c89713","type":"message","user":"UDDCM376D","text":"Promiseのunhanlded rejected調べてみます","ts":"1611272350.007000","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"6gm6"}]},{"client_msg_id":"b86397a0-0a4c-4b04-be8b-18cc7509eaf8","type":"message","user":"UDDCM376D","text":"お、動いたっぽい","ts":"1611272308.006500","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"YWuz"}]},{"client_msg_id":"50ef306e-a826-4857-8c5a-327bf6fc58d2","type":"message","user":"UAZ33BKV2","text":"そこが掴めないなら、まずJavaScriptのPromise関連の説明を沢山読んでからのほうがいい気がしないでもないけど...\nまあ動作テストして実動作に合せるのもありではある...あとで困らなければいいけど","ts":"1611272023.006300","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"RZZA"}]},{"client_msg_id":"394da435-0239-429e-a8cc-35e4ab8fa307","type":"message","user":"UDDCM376D","text":"ん、まず試してみるか","ts":"1611271858.005200","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"UIMem"}]},{"client_msg_id":"40b0f4ff-c8e1-49c8-82d9-deebaab77956","type":"message","user":"UDDCM376D","text":"そこ読みましたが、unhandled rejectedを使えば例外投げなくても良いですか？\n本文読んでもイマイチわからんかったんすよね","ts":"1611271849.004900","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"a5D"}]},{"client_msg_id":"d26057a4-a91b-4724-b202-7c47beb322c4","type":"message","user":"U03C4RC8X","text":"あれ、.catch()でキャッチできなかった感じなのかな","ts":"1611271161.003600","edited":{"user":"U03C4RC8X","ts":"1611271164.000000"},"team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"YLFQ"}]},{"client_msg_id":"44dc802e-58e6-490a-a7fe-3609745742de","type":"message","user":"UAZ33BKV2","text":"unhandled rejected のことね","ts":"1611271022.003200","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"aGVD"}]},{"client_msg_id":"7ae206b0-a1f8-4b16-809f-5b67a1a6aadf","type":"message","user":"UAZ33BKV2","text":"あのう\n\u003chttps://github.com/vim-jp/vital.vim/blob/621cda4f832d2f43f3b15cdbfb70ca36b219efc2/doc/vital/Async/Promise.txt#L416-L429\u003e\nのことではなく?\nこれもJavaScript Promise由来で、明示なthrow意外のVMの例外って、こういうのでcatchするしかないとおもうんだけど","ts":"1611270998.002900","attachments":[{"fallback":"GitHub: vim-jp/vital.vim","id":1,"title":"vim-jp/vital.vim","title_link":"https://github.com/vim-jp/vital.vim/blob/621cda4f832d2f43f3b15cdbfb70ca36b219efc2/doc/vital/Async/Promise.txt#L416-L429","text":"A comprehensive Vim utility functions for Vim plugins - vim-jp/vital.vim","thumb_url":"https://avatars.githubusercontent.com/u/1051958?s=400\u0026v=4","blocks":null,"service_name":"GitHub","service_icon":"https://a.slack-edge.com/80588/img/unfurl_icons/github.png","thumb_width":400,"thumb_height":400,"from_url":"https://github.com/vim-jp/vital.vim/blob/621cda4f832d2f43f3b15cdbfb70ca36b219efc2/doc/vital/Async/Promise.txt#L416-L429","original_url":"https://github.com/vim-jp/vital.vim/blob/621cda4f832d2f43f3b15cdbfb70ca36b219efc2/doc/vital/Async/Promise.txt#L416-L429"}],"team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"sr17"}]},{"client_msg_id":"c5a1ab80-ed75-45ad-baa2-b35ca88f908a","type":"message","user":"UDDCM376D","text":"ドキュメントを読んでいるとそうっぽいし、実際も動きもそうだけど、明示throwしなくてもキャッチしてくれる方法はないんかな","ts":"1611268097.000900","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"T7sI="}]},{"client_msg_id":"d7290066-1e0f-429d-b0fa-e753ad586362","type":"message","user":"UDDCM376D","text":"質問\nvitalのpromiseってthen内で明示的にthrowしないとエラーはキャッチしてくれないんすか？","ts":"1611268056.000200","team":"T03C4RC8V","replace_original":false,"delete_original":false,"blocks":[{"type":"rich_text","block_id":"dYG"}]}]
